### PR TITLE
Introduce finalize private field recipe

### PR DESCRIPTION
**Pull request by kunli2**
Issue: https://github.com/openrewrite/rewrite/issues/2611
Adds a recipe to add `final` modifier to private fields if the field is assigned only once somewhere in the class.
Rules: If a private field is assigned only once at the constructor, variable initializer, or initializer block, and there is no other assignment in the class, then it should be finalized.

### CODE REVIEW

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/FinalizePrivateFields.java [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+    }
+
+    private static class CollectPrivateFieldsAssignmentCounts extends JavaIsoVisitor<Map<JavaType.Variable, Integer>> {
+
+        /**
+         * Collects private fields assignment counts, count rules are:
+         * (1) any assignment in class constructor, instance variable initializer or initializer block count for 1.
+         * (2) any assignment in other place count for 2.
+         * (3) any assignment in a loop in anywhere count for 2.
+         * By giving 3 rules above, if a private field has an assigned count equal to 1, it should be finalized.
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/FinalizePrivateFields.java:119]: Clever!

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/FinalizePrivateFields.java [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+         *
+         * @param j The subtree to search, supposed to be at Class declaration level.
+         */
+        static void collect(J j, Map<JavaType.Variable, Integer> assignedCountMap) {
+            new CollectPrivateFieldsAssignmentCounts().visit(j, assignedCountMap);
+        }
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/FinalizePrivateFields.java:125]: `TreeVisitor.collect()` provides a shortcut/generalization of this pattern. I haven't checked that this compiles, but something like this:
```suggestion
        static Map<JavaType.Variable, Integer> collect(J j) {
             return TreeVisitor.collect(new CollectPrivateFieldsAssignmentCounts(), j, new HashMap<>());
        }
```

File: rewrite-java/src/main/java/org/openrewrite/java/cleanup/FinalizePrivateFields.java [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+import org.openrewrite.java.JavaIsoVisitor;
+import org.openrewrite.java.tree.Expression;
+import org.openrewrite.java.tree.J;
+import java.util.Map;
+import org.openrewrite.java.tree.JavaType;
+import org.openrewrite.java.tree.Space;
+import org.openrewrite.marker.Markers;
+
+
+public class FinalizePrivateFields extends Recipe {
```
Comment (sambsnyd) [inline rewrite-java/src/main/java/org/openrewrite/java/cleanup/FinalizePrivateFields.java:40]: Go ahead and add this to our [cleanup](https://github.com/openrewrite/rewrite/blob/0b994292372fce1f911ea7369a4aaf9afa2cb275/rewrite-java/src/main/resources/META-INF/rewrite/cleanup.yml#L18) and [common static analysis recipes](https://github.com/openrewrite/rewrite/blob/d408aaadddac717b00cc1f39e077bec84da51bc2/rewrite-java/src/main/resources/META-INF/rewrite/static-analysis-cleanup.yml#L18). Per Knut's helpful suggestion in the original issue, be sure it comes before the entry for RenamePrivateFieldsToCamelCase

**Review by sambsnyd — COMMENTED** [review]
Good work

**Comment by timtebeek** [comment]
Should this perhaps back off in the presence of Lombok annotations on the field or class? I'm thinking of `@Data,@Value,@Setter`...

**Comment by kunli2** [comment]
> 

Very good point! @timtebeek 
I tested with Lombok annotations and it compiles if apply `@Setter` to a final field (please see the code below), I saw somewhere said Lombok ignores all final fields, I think that's the reason.

```
import lombok.Data;
import lombok.Setter;
import lombok.Value;

@Value
@Data
class A {
    @Setter
    private final int a = 1;
}
```

**Comment by timtebeek** [comment]
Not sure I fully understand; I would expect the recipe not to make any fields final if the class or field is annotated with `@Data or @Setter`, as before the recipe run Lombok would then generate setter methods, whereas after the recipe has run, those setters are then no longer generated, but might still be used elsewhere. This could lead to compilation errors for a fairly common use case if not handled in the recipe. What are your views on this?

**Comment by kunli2** [comment]
Hi @timtebeek , thanks the explaining, this is the [PR](https://github.com/openrewrite/rewrite/pull/2688) to address your comment, please take a review when you got a chance, thanks

**Pull request merged**